File: app/src/main/java/be/digitalia/fosdem/utils/ThemeUtils.java
package be.digitalia.fosdem.utils;

import android.app.Activity;
import android.app.ActivityManager;
import android.content.Context;
import android.content.res.ColorStateList;
import android.graphics.ColorFilter;
import android.graphics.ColorMatrixColorFilter;
import android.graphics.drawable.Drawable;
import android.os.Build;
import android.util.TypedValue;
import android.view.View;
import android.widget.ImageView;

import androidx.annotation.ColorInt;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.graphics.drawable.DrawableCompat;
import be.digitalia.fosdem.R;

public class ThemeUtils {

	@SuppressWarnings("deprecation")
	public static void setActivityColors(@NonNull Activity activity, @ColorInt int taskColor, @ColorInt int statusBarColor) {
		if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
			activity.getWindow().setStatusBarColor(statusBarColor);
			final ActivityManager.TaskDescription taskDescription;
			if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
				taskDescription = new ActivityManager.TaskDescription(null, 0, taskColor | 0xFF000000);
			} else {
				taskDescription = new ActivityManager.TaskDescription(null, null, taskColor | 0xFF000000);
			}
			activity.setTaskDescription(taskDescription);
		}
	}

	public static void tintBackground(@NonNull View view, @Nullable ColorStateList backgroundColor) {
		final Drawable background = view.getBackground();
		if (background != null) {
			background.mutate();
			DrawableCompat.setTintList(background, backgroundColor);
		}
	}

	public static boolean isLightTheme(@NonNull Context context) {
		final TypedValue value = new TypedValue();
		return context.getTheme().resolveAttribute(R.attr.isLightTheme, value, true)
				&& value.data != 0;
	}

	public static void invertImageColors(@NonNull ImageView imageView) {
		final ColorFilter invertColorFilter = new ColorMatrixColorFilter(new float[]{
				-1.0f, 0.0f, 0.0f, 0.0f, 255.0f,
				0.0f, -1.0f, 0.0f, 0.0f, 255.0f,
				0.0f, 0.0f, -1.0f, 0.0f, 255.0f,
				0.0f, 0.0f, 0.0f, 1.0f, 0.0f
		});
		imageView.setColorFilter(invertColorFilter);
	}
}

File: app/src/main/java/be/digitalia/fosdem/utils/ThemeUtils.kt
package be.digitalia.fosdem.utils

import android.app.Activity
import android.app.ActivityManager.TaskDescription
import android.content.Context
import android.content.res.ColorStateList
import android.graphics.Color
import android.graphics.ColorFilter
import android.graphics.ColorMatrixColorFilter
import android.os.Build
import android.util.TypedValue
import android.view.View
import android.view.Window
import android.widget.ImageView
import androidx.annotation.ColorInt
import androidx.core.graphics.drawable.DrawableCompat

var Window.statusBarColorCompat: Int
    @ColorInt
    get() {
        return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) statusBarColor else Color.BLACK
    }
    set(@ColorInt color) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            statusBarColor = color
        }
    }

@Suppress("DEPRECATION")
fun Activity.setTaskColorPrimary(@ColorInt colorPrimary: Int) {
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
        val taskDescription = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
            TaskDescription(null, 0, colorPrimary or -0x1000000)
        } else {
            TaskDescription(null, null, colorPrimary or -0x1000000)
        }
        setTaskDescription(taskDescription)
    }
}

fun View.tintBackground(backgroundColor: ColorStateList?) {
    background?.let {
        it.mutate()
        DrawableCompat.setTintList(it, backgroundColor)
    }
}

val Context.isLightTheme: Boolean
    get() {
        val value = TypedValue()
        return theme.resolveAttribute(androidx.appcompat.R.attr.isLightTheme, value, true) && value.data != 0
    }

fun ImageView.invertImageColors() {
    val invertColorFilter: ColorFilter = ColorMatrixColorFilter(floatArrayOf(
            -1.0f, 0.0f, 0.0f, 0.0f, 255.0f,
            0.0f, -1.0f, 0.0f, 0.0f, 255.0f,
            0.0f, 0.0f, -1.0f, 0.0f, 255.0f,
            0.0f, 0.0f, 0.0f, 1.0f, 0.0f
    ))
    colorFilter = invertColorFilter
}